Support prefix term grammar for CONTAINS clause

jaspal007 opened the pull request:
Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to
support the grammar of prefix term to provide the intended result.

### 1. Issue:
- Fulltext search over `prefix_term` is not supported in Babelfish. When a `prefix_term` is passed as a search string in FTS TSQL CONTAINS query it throws `prefix term not supported error`
  - current output
     ```
     1> SELECT * FROM prefix_table WHERE CONTAINS(col, '"work*"')
     2> GO
    Msg 33557097, Level 16, State 1, Server persist-tanyagp-perf-diligent-16-6-i, Line 1
    Prefix term is not currently supported in Babelfish
     ```
### 2. Changes made to fix the issue:
- Grammar is added in FTS parser to handle whitespaces to access prefix_term token from the search string.  
- Custom function is created in the FTS parser to handle the prefix term token and handle all the cases of leading, trailing  and unwanted whitespaces within the `prefix_term` token. 

TASK: BABEL-4380
Authored-by: Jaspal Singh ijaspals@amazon.com
Signed-off-by: Jaspal Singh ijaspals@amazon.com


### Issues Resolved
BABEL-4380

### Test Scenarios Covered ###
* **Use case based -** YES


* **Boundary conditions -** YES


* **Arbitrary inputs -** YES


* **Negative test cases -** YES



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term
as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint
would throw unsupported error. To fix this, helper functions are created to
support the grammar of prefix term to provide the intended result.

TASK: BABEL-4380
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term
as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint
would throw unsupported error. To fix this, helper functions are created to
support the grammar of prefix term to provide the intended result.

TASK: BABEL-4380
Signed-off-by: Jaspal Singh ijaspals@amazon.com

jaspal007 committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL-4380-5_X

jaspal007 committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL-4380-5_X

roshan0708 commented on test/JDBC/input/full_text_search/fts-contains-prefix-vu-verify.sql:
```diff
@@ -0,0 +1,66 @@
+-- enable FULLTEXT
+-- tsql user=jdbc_user password=12345678
```
This is redundant. Please remove this at other places as well since it's a `.sql` file not `.mix`.

roshan0708 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,95 @@ static char
+
+    /* removing leading and trailing spaces outside of double quotes */
+    trim(trimmedInputStr, false);
+    inputLength = strlen(trimmedInputStr);
```
Seems redundant, we can remove this.

jaspal007 replied:
This function is being used to remove the extra spaces between the leading delimiters `"` and single quotes `'` and the trailing delimiters `"` and single quotes `'`
It handles cases when a customer might input search string with more than required spaces between single quotes and the delimiter `"`:
```
SELECT * FROM table_name WHERE CONTAINS({column_name | column_list}, '   "word1*"    ')
GO
```
This is treated as
```
SELECT * FROM table_name WHERE CONTAINS({column_name | column_list}, '"word1*"')
GO
```

thephantomthief replied:
do we need to do assignment here to `inputLength`? We are again overwriting it in L248

Commit:
Merge remote-tracking branch 'refs/remotes/origin/BABEL-4380-5_X' into BABEL-4380-5_X

roshan0708 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,95 @@ static char
+                errmsg("Special characters in the prefix term search condition are not currently supported in Babelfish")));
+        }
+        if (outputStr[i] == ' ') {
+            /* Removing multiple spaces and * from the search string */
+            while(i < inputLength-1 && (outputStr[i+1] == '*' || outputStr[i+1] == ' ')) {
+                i++;
+            }
+            if(output.len > 0){
+              if(i == inputLength-1) {
+                  appendStringInfoString(&output, ":*");
+              }
+              else {
+                  appendStringInfoString(&output, ":*<->");
+              }
+            } else {
+                  continue;
```
Better to mention example cases which are being handled here.

roshan0708 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,95 @@ static char
+                  continue;
+            }
+        } else if(outputStr[i] == '*') {
+                while(i < inputLength-1 && (outputStr[i+1] == '*' || outputStr[i+1] == ' ')) {
+                    i++;
+                }
+                /* adding the prefix operator ':*' and followed by operator '<->' if output is not NULL */
+                if(output.len > 0){
+                    if(i == inputLength-1) {
+                        appendStringInfoString(&output, ":*");
+                    }
+                    else {
+                        appendStringInfoString(&output, ":*<->");
+                    }
+                } else {
+                        continue;
+                }
```
Please mention example cases here also.

roshan0708 commented on test/JDBC/jdbc_schedule:
```diff
@@ -10,6 +10,11 @@
 
 all
 
+# Ignore fts-contains-before files as they don't have test cases that support prefix term support
+ignore#!#fts-contains-before-vu-prepare
+ignore#!#fts-contains-before-vu-verify
+ignore#!#fts-contains-before-vu-cleanup
+
```
Seems redundant, we can altogether remove this if not needed.

roshan0708 commented on test/JDBC/input/full_text_search/fts-contains-prefix-vu-verify.sql:
```diff
@@ -0,0 +1,66 @@
+GO
+
+-- Sample search query using CONTAINS
+-- 1. Basic prefix search across all text fields
```
Add more tests cases around different patterns of prefix terms as mentioned in the test plan:
- "one two three*"
- "one* two three*"
- "one* two* three*"
- "one*two*three*"
- other combinations with spaces and `*` characters

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-4380-5_X

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

TASK: BABEL-4380
Signed-off-by: Jaspal Singh ijaspals@amazon.com

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,109 @@ static char
+
+
+    /* Check for empty input - this should not be possible based on lexer rules, but check just in case */
+    if (!inputStr || !(inputLength = strlen(inputStr))) {
```
do we need to do assignment here to `inputLength`?

roshan0708 replied:
We can just simply check `(!inputStr || !*inputStr)`

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,109 @@ static char
+    initStringInfo(&output);
+
+    for(int i=0; i<inputLength; i++) {
+        if (strchr(specialChars, outputStr[i]) != NULL || strchr("`'_", outputStr[i]) != NULL) {
```
why don't we put all the special chars in `specialChars`

roshan0708 replied:
for strings with special characters "`'_", these result in exact matches for simple term but @jaspal007 We should check if such prefix term search conditions match the same results, then only we should use this.

roshan0708 replied:
@jaspal007 Current implementation of this function can be further simplified. IMO:
* Simplify nested conditions for spaces/asterisks handling into single pattern
* Reduce memory buffers from two to one for better memory management
* Consolidate redundant string operations for `':*'` and `'<->'` separators
* Use pointer arithmetic instead of array indexing for better performance

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,109 @@ static char
+                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                errmsg("Special characters in the prefix term search condition are not currently supported in Babelfish")));
+        }
+        if (outputStr[i] == ' ') {
```
what about tab, new line characters?

jaspal007 replied:
The intended output of sql server is that it returns no effective records if we have tab or newline character present in prefix term search string.

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

TASK: BABEL-4380
Signed-off-by: Jaspal Singh ijaspals@amazon.com

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,109 @@ static char
+             * Case 1: '"word1   * * ** *"' = 'word1:*'
+             * Case 2: '"word1   * * ** * word2*"' = 'word1:*<->word2:*'
+             */
+            while(i < inputLength-1 && (outputStr[i+1] == '*' || outputStr[i+1] == ' ')) {
```
what about tab, new line characters?

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,108 @@ static char
+            } else {
+                  continue;
+            }
+        } else if(outputStr[i] == '*') {
```
Can we combine the above else if with this block

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,108 @@ static char
+                appendStringInfoChar(&output, outputStr[i]);
+        }
+    }
+    appendStringInfoChar(&output, '\0');
```
This is not needed

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-4380-5_X

roshan0708 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,93 @@ static char
+     * Rewriting search string in format word1:*<->word2:*  
+     */
+    while(*trimmedOutputStr) {
+        if (strchr(specialChars, *trimmedOutputStr) != NULL) {
+            ereport(ERROR,
+                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                errmsg("Special characters in the prefix term search condition are not currently supported in Babelfish")));
+        }
```
When an error is thrown here, both `outputStr` and `output.data` memory will be leaked. Please free these allocations before calling ereport().

roshan0708 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,93 @@ static char
+                 * When FTS is performed over prefix_term,
+                 * it outputs no result
+                 */ 
+                pfree(outputStr);
+                output.data = "";
```
setting `output.data` to a string literal after initialization without proper cleanup will cause memory leaks. Use `resetStringInfo(&output)` instead to properly clean up the previously allocated memory.

roshan0708 commented:
We should add test cases for emojis, CTEs as well

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-4380-5_X

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-4380-5_X

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,141 @@ static char
+     * trailing single quote (') and trailing delimiter (")
+     * '   "word1 word2*"  ' = '"word1 word2*"'
+     */
+    trim(output, false);
```
doesn't `trim(output, true)` not also do this implicitly?

jaspal007 replied:
When we call `trim(output, true)` it processes the leading spaces within the double quotes and search string.
To remove the trailing spaces outside double quotes, we implement `trim(output, false)`

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,141 @@ static char
+    while(start<=end) {
+        if (strchr(specialChars, *start) != NULL) {
+            pfree(output);
+            resetStringInfo(&outputStr);
```
isn't this already empty at this point?

jaspal007 replied:
This is for the case if special characters are present in the middle of search stirng
`'"word@1 wor$d2*"'`

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,141 @@ static char
+        * Case 3: '"word1* *' + CHAR(9) + '** *"' = 'word1:*'
+        * Case 4: '"word1* * *  ' + CHAR(9) + '* ' + CHAR(9) + ' *** * word2*"' = 'word1:*<->word2:*'
+        */
+        if(*start == ' ' || *start == '*' || *start == '\t') {
```
if there is a newline character, we shouldn't remove the following spaces, tabs etc.?

jaspal007 replied:
That is handled in `if` clause for newline character separately

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,141 @@ static char
+         * a newline character 
+         * at the end of the search string
+         */
+        while(start<=end && (*end == ' ' || *end == '*' || *end == '\t' || *end == '\n')) {
```
we only need to run this loop once so we can take it out of the outer while loop

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,141 @@ static char
+             */
+            if(*(start + 1) == '\n') {
+                start++;
+                continue;
```
why do we need `continue` here?

jaspal007 replied:
we require `continue` here as the newline character is handled in a separate `if` clause

thephantomthief commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,141 @@ static char
+                    char *newlineHash = replace_special_chars_fts_impl("\n");
+                    trim(newlineHash, false);
+                    appendStringInfoString(&outputStr, ":*<->");
+                    appendStringInfo(&outputStr, "%s", newlineHash);
```
why not use `appendStringInfoString` again here?

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -2187,21 +2188,23 @@ char
-		unique_hashes[i] = construct_unique_index_name("specialChars", psprintf("cat%d", i + 1));
+	for (int i = 0; i < 5; i++) {
+		unique_hashes[i] = construct_unique_index_name("ftSpecial", psprintf("cat%d", i + 1));
+		unique_hashes[i][21] = '\0';
```
Are we certain every unique hash will at least be 22 chars long?

jaspal007 replied:
the unique hash value generated by `construct_unique_index_name` is quite large, we require a hash value that is smaller and yet has a good collision avoidance

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

TASK: BABEL-4380
Signed-off-by: Jaspal Singh ijaspals@amazon.com

roshan0708 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,142 @@ static char
+    leftPtr = output;
+    rightPtr = output + (strlen(output) - 1);
+        
+    /*
+        * trim the extra 
+        * space, 
+        * asterisk, 
+        * a tab character or 
+        * a newline character 
+        * at the end of the search string
+        */
```
nit: Indentation. Also, we can condense this comment in a line or two.

roshan0708 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -195,6 +205,142 @@ static char
+        * Case 3: '"word1* *' + CHAR(9) + '** *"' = 'word1:*'
+        * Case 4: '"word1* * *  ' + CHAR(9) + '* ' + CHAR(9) + ' *** * word2*"' = 'word1:*<->word2:*'
+        */
+        if(*leftPtr == ' ' || *leftPtr == '*' || *leftPtr == '\t') {
```
nit: Spacing after `if`. Check other places as well.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -50,6 +50,7 @@ extern bool restore_tsql_tabletype;
 extern bool babelfish_dump_restore;
 extern char *get_cur_db_name(void);
 extern char *construct_unique_index_name(char *index_name, char *relation_name); 
+extern char *generate_fts_char_hash(char *prefix, char *bucket);
```
We can remove this now since it's redundant.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -2200,8 +2201,9 @@ char
-		for (int j = 0; j < 4; j++) {
-			if (strchr(special_chars[j], input_str[i]) != NULL) {
+		for (int j = 0; j < 5; j++) {
+			char *chr;
+			if ((chr = strchr(special_chars[j], input_str[i])) != NULL) {
```
What is this change for? Are we using `chr` somewhere else?

roshan0708 commented:
Regarding tests, I think creating three separate tables for testing is redundant and we could write tests using a single table itself.

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

TASK: BABEL-4380
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Support prefix term grammar for CONTAINS clause

Previously in Babelfish, the feature of fulltext search for prefix term as search string in CONTAINS clause was not supported. Prefix term over TDS endpoint would throw unsupported error. To fix this, helper functions are created to support the grammar of prefix term to provide the intended result.

TASK: BABEL-4380
Signed-off-by: Jaspal Singh ijaspals@amazon.com

jaspal007 commented:
nit: fix spacings and indentations

jsudrik reviewed (approved):
Approved.
